### ISSUE TITLE
Slow polynomial factoring

### ISSUE THREAD

**Issue by wbhart** [wbhart/flint2#932]
When I implemented polynomial factoring in 2016, the following were the times:

Poly           NTL             Magma           Flint-1.6                New
==================================================
P1             0.07               0.03               0.10                    0.30
P2              0.2                0.08               0.10                    0.50
P3               0.3               0.16               0.16                    0.93
P4               2.0                1.9                0.94                     4.6
P5              0.09               0.11               0.04                    0.03
P6              0.12               0.11               0.11                    0.26
P7                1.1               1.1                0.52                     1.8
P8                3.4               2.2                 1.5                    0.79
M12_5         12.4               9.5                 2.9                     2.2
M12_6         21.7              21.5                5.2                      73
S7               0.34              0.42               0.20                  0.56
S8                3.8                4.6                2.1                     7.9
S9                 71               165                 21                      ----
S10               ??                 ??                 ??                      ----
T1                 3.8               2.5                 1.2                     0.6
T2                 3.2               2.1                 1.2                     0.6
T3                  24                20                 7.4                      59 
H1                  ??                ??                 ??                      9.3
H2                  ??                ??                 ??                      ----

I also gave a list of things that we don't do yet, some of which have been done since:

* tuning of parameters p^a, rho, delta, eta, l, E_bound, number of CLDs, etc a la van Hoeij-Novocin
* switch to Zassenhaus once number of local factors is low enough
* inflation/deflation (aka power hack)
* U_LLL floating point LLL trick
* spend more time searching for optimal prime p
* mix CLD data using random matrices
* only run LLL if justified as outlined in van Hoeij-Novocin
* only add row if justified as outlined in van Hoeij-Novocin
* speed up basic arithmetic in LLL
* fast divisibility testing, e.g. mod a small prime
* asymptotically fast truncated division smod p^a

**Comment by fredrik-johansson** [comment]
We've made definite progress in the last year.

Here's the improvement over our last release on my machine, with #930 and #937 merged:

             Flint-2.7.1   Current
    P1         143 ms       16 ms     (8.9x faster)
    P2         275 ms       33 ms     (8.3x faster)
    P3:        512 ms       65 ms     (7.9x faster)
    P4:       2325 ms      579 ms     (4.0x faster)
    P5:         17 ms       19 ms     (0.9x faster)
    P6:        125 ms       28 ms     (4.5x faster)
    P7:        925 ms      649 ms     (1.4x faster)
    P8:        341 ms      332 ms     (1.0x faster)
    M12_5:    1173 ms      996 ms     (1.2x faster)
    M12_6:   37056 ms    15905 ms     (2.3x faster)
    T1:        317 ms      297 ms     (1.1x faster)
    T2:        350 ms      315 ms     (1.1x faster)
    T3:      26561 ms     9832 ms     (2.7x faster)
    H1:       4521 ms     3381 ms     (1.3x faster)
    S7:        278 ms      260 ms     (1.1x faster)
    S8:       4284 ms     3686 ms     (1.2x faster)

S9, S10 and H2 still take forever.

(Note: #937 also adds the new test polynomial C1 from #907 to the test suite. This takes 680 ms with the latest improvements.)

**Comment by fredrik-johansson** [comment]
Correction: S9 doesn't take forever, it only takes 12 minutes!

To get a sense of what is going on, I put a printout in factor_van_hoeij to show each call to LLL and how long it takes:

S5:

    LLL 17 x 17, 35 bits    : 0.00021 s
    LLL 17 x 18, 35 bits    : 0.000162 s
    LLL 17 x 19, -35 bits    : 0.000411 s
    LLL 10 x 20, -35 bits    : 0.000158 s
    LLL 8 x 21, -35 bits    : 0.000217 s

S6:

    LLL 17 x 17, 35 bits    : 0.000198 s
    LLL 17 x 18, -35 bits    : 0.000843 s
    LLL 18 x 19, -35 bits    : 0.000288 s
    LLL 18 x 20, -35 bits    : 0.001144 s
    LLL 33 x 33, 54 bits    : 0.00122 s
    LLL 33 x 34, -54 bits    : 0.001891 s
    LLL 30 x 35, -54 bits    : 0.002212 s
    LLL 31 x 36, -54 bits    : 0.001956 s
    LLL 15 x 37, -54 bits    : 0.000606 s
    LLL 5 x 38, -54 bits    : 0.00014 s

S7:

    LLL 33 x 33, 54 bits    : 0.000697 s
    LLL 33 x 34, -54 bits    : 0.001945 s
    LLL 24 x 35, -54 bits    : 0.00034 s
    LLL 24 x 36, -54 bits    : 0.002197 s
    LLL 25 x 37, -54 bits    : 0.000594 s
    LLL 25 x 38, -54 bits    : 0.002948 s
    LLL 65 x 65, 103 bits    : 0.004933 s
    LLL 65 x 66, -103 bits    : 0.031963 s
    LLL 66 x 67, -103 bits    : 0.00867 s
    LLL 62 x 68, -103 bits    : 0.01468 s
    LLL 63 x 69, -103 bits    : 0.092595 s
    LLL 31 x 70, -103 bits    : 0.069487 s
    LLL 30 x 71, -103 bits    : 0.010149 s

S8:

    LLL 65 x 65, 103 bits    : 0.004821 s
    LLL 65 x 66, -103 bits    : 0.030986 s
    LLL 66 x 67, -103 bits    : 0.050655 s
    LLL 54 x 68, -103 bits    : 0.050977 s
    LLL 55 x 69, -103 bits    : 0.053301 s
    LLL 129 x 129, 200 bits    : 0.084153 s
    LLL 129 x 130, -200 bits    : 0.084978 s
    LLL 125 x 131, -200 bits    : 0.402601 s
    LLL 126 x 132, -200 bits    : 0.132055 s
    LLL 114 x 133, -200 bits    : 0.145725 s
    LLL 115 x 134, -200 bits    : 1.60024 s
    LLL 54 x 135, -200 bits    : 0.79608 s
    LLL 54 x 136, -200 bits    : 0.181834 s

S9:

    LLL 129 x 129, 200 bits    : 0.082886 s
    LLL 129 x 130, -200 bits    : 0.21772 s
    LLL 130 x 131, -200 bits    : 0.435493 s
    LLL 114 x 132, -200 bits    : 0.759173 s
    LLL 115 x 133, -200 bits    : 1.62082 s
    LLL 116 x 134, -200 bits    : 1.38855 s
    LLL 117 x 135, -200 bits    : 1.49151 s
    LLL 257 x 257, 393 bits    : 0.670866 s
    LLL 257 x 258, -393 bits    : 3.12472 s
    LLL 252 x 259, -393 bits    : 3.86854 s
    LLL 253 x 260, -393 bits    : 91.2537 s
    LLL 237 x 261, -393 bits    : 1.93611 s
    LLL 238 x 262, -393 bits    : 7.50094 s
    LLL 198 x 263, -393 bits    : 303.488 s
    LLL 199 x 264, -393 bits    : 335.692 s
    LLL 86 x 265, -393 bits    : 2.05302 s

(This is with MPFR disabled.)

It looks like LLL is taking a long time to finish for some input.

**Comment by wbhart** [comment]
My best guess is that LLL should only be called if it is worth it. Flint1 had a function to determine if this was the case. I'm quite surprised this made so much difference if it was the real reason for the better performance there.

**Comment by fredrik-johansson** [comment]
Playing with LLL parameters makes some difference (up to a factor two):

Note that the defaults are delta = 0.99, eta = 0.51.

delta = 0.75, eta = 0.51

    LLL 129 x 129, 200 bits    : 0.085197 s
    LLL 129 x 130, -200 bits    : 0.118409 s
    LLL 130 x 131, -200 bits    : 0.3499 s
    LLL 114 x 132, -200 bits    : 0.415874 s
    LLL 115 x 133, -200 bits    : 0.693953 s
    LLL 116 x 134, -200 bits    : 0.915278 s
    LLL 117 x 135, -200 bits    : 0.845167 s
    LLL 257 x 257, 393 bits    : 0.692637 s
    LLL 257 x 258, -393 bits    : 3.23055 s
    LLL 252 x 259, -393 bits    : 1.78256 s
    LLL 253 x 260, -393 bits    : 60.4798 s
    LLL 237 x 261, -393 bits    : 1.25693 s
    LLL 238 x 262, -393 bits    : 2.3563 s
    LLL 239 x 263, -393 bits    : 240.758 s
    LLL 240 x 264, -393 bits    : 253.835 s
    LLL 127 x 265, -393 bits    : 3.27139 s

delta = 0.75, eta = 0.6

    LLL 129 x 129, 200 bits    : 0.089295 s
    LLL 129 x 130, -200 bits    : 0.113068 s
    LLL 130 x 131, -200 bits    : 0.332855 s
    LLL 114 x 132, -200 bits    : 0.284462 s
    LLL 115 x 133, -200 bits    : 0.546179 s
    LLL 116 x 134, -200 bits    : 0.611776 s
    LLL 117 x 135, -200 bits    : 0.753995 s
    LLL 118 x 136, -200 bits    : 0.67701 s
    LLL 257 x 257, 393 bits    : 0.765197 s
    LLL 257 x 258, -393 bits    : 3.47659 s
    LLL 252 x 259, -393 bits    : 1.47646 s
    LLL 253 x 260, -393 bits    : 56.906 s
    LLL 237 x 261, -393 bits    : 0.821933 s
    LLL 238 x 262, -393 bits    : 2.22039 s
    LLL 239 x 263, -393 bits    : 182.26 s
    LLL 240 x 264, -393 bits    : 196.611 s
    LLL 127 x 265, -393 bits    : 2.63544 s

delta = 0.75, eta = 0.75

    LLL 129 x 129, 200 bits    : 0.088646 s
    LLL 129 x 130, -200 bits    : 0.081249 s
    LLL 130 x 131, -200 bits    : 0.297023 s
    LLL 114 x 132, -200 bits    : 0.178078 s
    LLL 115 x 133, -200 bits    : 0.327666 s
    LLL 116 x 134, -200 bits    : 0.454545 s
    LLL 117 x 135, -200 bits    : 23.5608 s
    LLL 117 x 136, -200 bits    : 0.535455 s
    LLL 257 x 257, 393 bits    : 0.68268 s
    LLL 257 x 258, -393 bits    : 3.2413 s
    LLL 252 x 259, -393 bits    : 1.0691 s
    LLL 253 x 260, -393 bits    : 42.3917 s
    LLL 237 x 261, -393 bits    : 0.670725 s
    LLL 238 x 262, -393 bits    : 1.11835 s
    LLL 239 x 263, -393 bits    : 144.582 s
    LLL 240 x 264, -393 bits    : 145.512 s
    LLL 239 x 265, -393 bits    : 2.33037 s
    LLL 240 x 266, -393 bits    : 5.09456 s

**Comment by wbhart** [comment]
Sure but one has to use parameters within the scope of Stehle's paper. Of
course there's nothing stopping one doing a final reduction with other
values, but this might be more expensive, as one then has to use a standard
LLL reduction.

**Comment by fredrik-johansson** [comment]
Results with

* speed up basic arithmetic in LLL

done (though more LLL speedups are still possible):

| Polynomial | Degree | Bits | Local factors | Factors | FLINT 3.3 | FLINT 3.4-dev | Speedup |
|------------|--------|------|---------------|---------|-----------|---------------|---------|
| P1         | 156    | 1407 | 10            | 36      | 0.007 s   | 0.007 s       |         |
| P2         | 196    | 1391 | 7             | 12      | 0.011 s   | 0.011 s       |         |
| P3         | 336    | 1982 | 7             | 16      | 0.022 s   | 0.022 s       |         |
| P4         | 462    | 2511 | 42            | 2       | 0.159 s   | 0.158 s       |         |
| P5         | 64     | 131  | 32            | 1       | 0.007 s   | 0.007 s       |         |
| P6         | 144    | 548  | 24            | 6       | 0.010 s   | 0.010 s       |         |
| P7         | 384    | 187  | 88            | 1       | 0.382 s   | 0.218 s       | 1.50x   |
| P8         | 972    | 707  | 54            | 1       | 0.120 s   | 0.114 s       |         |
| M12_5      | 792    | 3623 | 72            | 1       | 0.376 s   | 0.328 s       | 1.14x   |
| M12_6      | 924    | 4870 | 84            | 2       | 4.920 s   | 4.851 s       |         |
| T1         | 900    | 915  | 30            | 2       | 0.102 s   | 0.102 s       |         |
| T2         | 900    | 421  | 32            | 2       | 0.108 s   | 0.110 s       |         |
| T3         | 2401   | 1541 | 63            | 4       | 1.909 s   | 1.768 s       | 1.08x   |
| P7*M12_5   | 1176   | 3787 | 188           | 2       | 278 s     | 62.4 s        | 4.45x   |
| P7*M12_6   | 1308   | 5029 | 206           | 3       | 535 s     | 120 s         | 4.47x   |
| H1         | 960    | 955  | 131           | 28      | 1.927 s   | 0.935 s       | 2.06x   |
| S7         | 128    | 289  | 64            | 1       | 0.134 s   | 0.088 s       | 1.52x   |
| S8         | 256    | 624  | 128           | 1       | 1.431 s   | 1.077 s       | 1.32x   |
| C1         | 1024   | 478  | 64            | 32      | 0.556 s   | 0.256 s       | 2.17x   |
| H2         | 4096   | 2    | 256           | 6       | 443.8 s   | 86.7 s        | 5.12x   |
| S9         | 512    | 1334 | 256           | 1       | 371.8 s   | 70.7 s        | 5.25x   |
| S10        | 1024   | 2837 | 512           | 1       | 28548 s   | 6994 s        | 4.08x   |


S9 is still slower than FLINT 1. Which optimizations listed in the OP are still missing?

**Comment by fredrik-johansson** [comment]
As far as I can tell, after #2698, the only tricks from Bill's list still missing are the following:

* tuning of parameters p^a, rho, delta, eta, l, E_bound, number of CLDs, etc a la van Hoeij-Novocin

Various parameters differ betweent the current code and FLINT 1.6. That doesn't mean the original parameters are necessarily optimal; they would need to be retuned for FLINT 3 arithmetic.

* spend more time searching for optimal prime p

The current heuristic looks identical to FLINT 1.6 (try first three good primes) and looks pretty reasonable. Of course we could try experimenting with different methods.

* mix CLD data using random matrices
* only run LLL if justified as outlined in van Hoeij-Novocin
* only add row if justified as outlined in van Hoeij-Novocin

These are potentially significant. I tried to implement these with AI assistance and got either no change in performance or much worse performance, but that may just be the AI's fault. Pari still factors Swinnerton-Dyer polynomials a bit faster than FLINT, so something is clearly missing.

* asymptotically fast truncated division smod p^a

`fmpz_poly_divhigh_smodp` and `fmpz_poly_divlow_smodp` still use basecase division only, so this should be a straightforward improvement. Not sure how often this is a bottleneck. Would be nice to have these on top of a general-purpose signed version of an `fmpz_mod` base ring, not just ad-hoc for the factoring code.

**Comment by wbhart** [comment]
From memory, tuning was quite critical if you care about Swinnerton-Dyer polynomials. There's a big step up to each new one, and if you didn't target the largest one you cared about, it could easily be very slow.

I remember Andy tried a bunch of stuff to speed those up, including mixing the LLL matrix with random operations, to distribute the data more uniformly across the rows. I think the idea was to try and reduce the size of the arithmetic needed. I don't recall if he ever got that to work.

**Comment by wbhart** [comment]
What is the actual bottleneck in factoring SD polys? From memory it is proving that the SD poly is irreducible.

The SD polynomials are even, i.e. f(x) = g(x^2). So we can apply van Hoeij-Novocin to g instead of f. I vaguely recall we do this. This is absolutely required for good timings. (One has to use Capelli's condition to check for irreducibility of f. I don't remember if we do this exactly.)

Van Hoeij-Novocin basically lifts to a small precision, say p^3 first, then uses van Hoeij to detect linear relations between modular factors, essentially determining which modular factors must "travel together" in the final factorisations (to use the terminology Andy used to use). This effectively shrinks the lattice. I was never convinced that the code for keeping track of these linear relations between modular factors was very efficient. Claude might be able to come up with something better.

To avoid redundancy when adding traces to the matrix, it only does this if the rank of the matrix increases. This is critical for the SD case, as the Galois group has the form (C_2)^k and adding additional traces is often redundant. I wonder if this rank computation is some kind of bottleneck? (Actually, it sounds from the above like we don't even do this test, which is almost certainly required for fast SD factorisation. Perhaps it turned out to be harder to get efficient than original thought, due to a poor quality rank test?)

There may be other tricks, such as using statistical bounds for the traces. Once you know that polynomials factor into linear and quadratic factors modulo every prime tested, you know something about the statistical distribution of the roots (including of the factors), e.g. likely normally distributed. For the LM bound (which is not used in vHN) at least, this can be used in practice to come up with an easy to compute statistical bound. In the vHN case, one has an expectation for the value of the traces which is strictly positive. Apparently one can shift the lattice to move this expected value to zero, reducing the bound needed when searching for factors. (Citation needed.)

From memory, we currently switch from hunting for factors to proving irreducibility once we stop finding candidate factors. This prevents us from lifting to too high a precision p^k when hunting for quick factors. If the bottleneck is in this phase, then probably the bound is too high, and maybe we should use some statistical test to identify we are likely in a SD-like case, and prefer a lower bound before switching to proving irreducibility, as just outlined. Perhaps PariGP does something like this.

I guess we should check if PariGP is detecting SD-like polynomials and using additional tricks like this to minimise the time, knowing that SD polynomials are a standard benchmark test.

**Comment by fredrik-johansson** [comment]
The bottleneck is running LLL; I guess Pari's setup somehow generates fewer LLL calls or smaller coefficients.

I have thought about detecting SD-like polynomials, for some definition of "SD-like". I think you could use LLL to guess a basis of small integers $a_i$ such that $\sum \sqrt{a_i}$ is a root of the polynomial, and then verify this guess and check minimality. This has a decent chance of being faster for SD polys, but might not be worth slowing down generic input unless there is some really fast way to screen out inputs that aren't actually in this class.

**Comment by wbhart** [comment]
That seems very specific to SD. I'd say you want a test to check for roots that are normally distributed. I think that can be done by checking factorisation modulo a number of primes p. As the factorisation modulo p reflects the structure of the Galois group by the Chebotarev density theorem, it turns out that you'll only get linear and quadratic factors for SD-like polynomials.

Once you know that, the Galois group is an elementary 2-group, and so by Kummer theory, the field is generated by square roots. If the polynomial is to be irreducible, a primitive element is usually given as a sum of such square roots. The elements of the Galois group will just flip the signs and so the roots will end up being normally distributed around the origin in the complex plane (or the real line, in the case of SD polynomials).

So this is a reasonable heuristic test for SD polynomials. Moreover, we already look at the polynomial factorisation modulo multiple primes, and it's extremely unlikely to have this linear/quadratic factorisation modulo every such prime unless we have an SD-like polynomial.

**Comment by wbhart** [comment]
But I don't think we should just outright cheat and check for SD polynomials. We should use the fact that the polynomial is SD-like to speed up vHN.

The only hope I see for that is in switching earlier to the irreducibility test, based on some statistical approximate bound based on the normal distribution of the roots (rather than basing it on the size of the largest possible root, for example).

**Comment by wbhart** [comment]
So perhaps using the rank test as per vHN is useful for SD-like polynomials. If we aren't doing that, I'm fairly sure this will be the problem, as adding more traces (CLDs) is highly redundant for SD polynomials. They should only be added if the rank increases.

**Comment by wbhart** [comment]
Apparently you can also use the FFT to generate CLDs (citation needed). This is very fast. Of course, one needs the rank test before just adding them to the matrix.

I don't recall us using the FFT to generate them though, so this may be an improvement to consider.

**Comment by wbhart** [comment]
Actually, just naively checking the rank is probably inefficient. You only want to know if the rank has increased when you add more CLDs. I'm sure you can store the original matrix in some form which makes it easy to check the rank has increased when the new columns are added.

Quite how that works with increasing precision probably requires some thought. But it seems like it ought to be possible to save some time even when the precision increases.

**Comment by fredrik-johansson** [comment]
Current timings vs Pari and NTL, excluding the slow S10. Obtained on a very noisy system, and the Pari and NTL installs are possibly out of date or not configured optimally, so take this with a grain of salt, but it should give some idea.

```
Polynomial    deg   PARI (s)   NTL (s)  FLINT (s)
--------------------------------------------------
P1            156      0.010     0.017      0.005
P2            196      0.030     0.024      0.011
P3            336      0.024     0.044      0.023
P4            462      0.298     0.239      0.140
P5             64      0.007     0.016      0.006
P6            144      0.010     0.021      0.010
P7            384      0.164     0.320      0.149
P8            972      0.601     0.390      0.090
M12_5         792      1.322     3.115      0.438
M12_6         924      2.435     5.812      3.630
T1            900      0.737     0.303      0.084
T2            900      0.593     0.253      0.090
T3           2401      5.660     2.032      1.820
H1            960      0.945     1.311      1.139
S7            128      0.056     0.121      0.048
S8            256      0.716     1.540      0.936
C1           1024      0.086     0.185      0.145
S9            512     28.328    25.718     28.620
H2           4096      3.526     6.027     18.223
P7*M12_5     1176     57.207   237.007     30.641
P7*M12_6     1308     81.806   366.481     52.014
```

The biggest underperformer is now H2 by a large margin; I wonder what's going on there.

**Comment by wbhart** [comment]
I had a look through my email to see if Andy had specifically commented on H2. I didn't find anything specific, but he did mention asymptotically fast division being important for some of these factorisations. Unfortunately, I don't have any specific recollections for H2. I assume the bottleneck is mainly in local factoring, as I can find previous messages where you did work on that (which Andy commented on)? However, all the timings looked good against NTL back then (admittedly a long time ago).

**Comment by fredrik-johansson** [comment]
According to `perf` most time on H2 is spent in LLL (or subroutines thereof).

**Comment by wbhart** [comment]
I think H2 is maybe $\Phi_{255}(x^{32})$. It has 6 factors with degrees 128, 128, 256, 512, 1024, 2048.

The coefficient height is very small and so lifting bounds are tiny.

Some implementations do the following. They spend some time finding a good prime, to maximise the size of the local factors. Though pretty much modulo any prime there will be a large number of local factors for H2.

Then they start with few traces (maybe even just one) and hopefully reduce the LLL dimension a lot.

Once the dimension gets below about 40, they use meet-in-the-middle subset sum tests with random linear combinations of the traces, to partition the problem. (A true factor has traces which are integers in a small range, and so a small linear combination of them is still a small integer. Subset sum tests can check if certain combinations of values mod p^k can ever be in this small range.) This saves having to run LLL on larger dimension at all, by partitioning the problem.

If Pari is doing less LLL, perhaps it is doing this.

Like everything, there is a juggling act to ensure running subset sums doesn't blow out the timings for cases vHN handles effortlessly.

**Comment by wbhart** [comment]
From memory, NTL has a very efficient subset sum test, used by its Zassenhaus. Flint's one isn't as good, unless someone improved it at some point. I don't know about Pari.

**Comment by wbhart** [comment]
Asymptotically fast division might be an issue with raising the bound for switch to subset sum, as presumably on polynomials so large, there are some polynomial divisions which are quite large.

I vaguely recall that when Andy first implemented vHN, Flint did Zassenhaus up to some number of local factors (10?) then went all or nothing on vHN.

Anyway, being able to switch back to subset sum would definitely save LLL calls, which seems feasibly what is going wrong with H2. But Flint doesn't work this way at all (see below).

It seems we didn't have H2 times for Flint-1.6 for some reason. I wonder if I can find some old timings somewhere.

**Comment by wbhart** [comment]
No, it looks like the timings at the top of this issue are from Oct 2016, at which time I did not have timings for H2 either for the old Flint or the new Flint implementations. There was no H2 in the ISSAC paper either.

At that time, it looks like the new implementation couldn't even factor H2. It isn't clear to me if Flint 1.6 could do it either. So any progress from here is probably going to require careful optimisation, but should be worthwhile.

**Comment by wbhart** [comment]
Actually, looking through the flint-3.2 implementation (just what I have installed at present), it doesn't switch to subset-sum at all. Instead it relies entirely on removal. As it knows that true factors correspond to 0-1 combinations, it can aggressively remove vectors when the Gram-Schmidt bound (gs_B) is exceeded by any vector, as they cannot possibly contribute to a true factor. (Any true factor will correspond to a 0-1 vector, with 1's indicating which local factor is part of that global factor. As the length of any 0-1 matrix can't have GS norm greater than r, the number of local factors, and as LLL is always combining integer combinations of rows, any row whose GS norm already exceeds r cannot possibly ever contribute to a global factor and can be aggressively removed.)

So the number of rows of the LLL matrix is being reduced all the time.

So Flint is relying on this, instead of switching back to subset sum (or on reducing the number of columns; see below).

I wonder which approach Pari/GP takes.

We do a quick mod 2 divisibility test to save trial divisions, so maybe that is why this is not a bottleneck.

**Comment by wbhart** [comment]
There are two things that can be done to reduce dimension:

Firstly, in algorithms that wait for 0-1 vectors to perfectly partition the local factors, instead of looking for a perfect 0-1 partition, they also scan for cases where there is a single 0-1 vector (that is not the all 1's vector), pushed to the top by LLL. If one occurs, it could correspond to a factor of the original polynomial. It can be quickly tested with the mod 2 factorisation and then a trial division. If a true global factor is found this way, the local factors can be removed, some stuff recomputed and the algorithm can resume with smaller LLL dimension.

Secondly, if two colums end up being identical, these local factors travel together, and so the local factors can be glued, reducing the dimension. However, one has to be careful about gluing before the precision is high enough. One has to either prove the precision is high enough that identical rows do belong together, or one has to entirely trash the matrix (and gluing) and start again at the higher precision, if one glued before reaching this proven precision bound. So this trick is actually not that easy to implement.

Flint does do a check if identical columns, when glued, produce real factors. But it takes an all or nothing approach. Unless the polynomial fully factors, it just throws away the global factors found and continues. It does not in practice actually do any gluing to reduce dimension!

In particular, Flint does not remove global factors, rebuild and start again.

The main idea here is that if you have enough traces and high enough precision to start identifying true factors, it is very likely that only one more LLL pass will be required to finish off entirely. Rebuilding will take far too long. As LLL works on rows, one can't even efficiently virtually combine the columns to save duplicate arithmetic in LLL. If you do combine columns (i.e. combine the corresponding local factors to reduce dimension), you have to recompute the traces, reallocate the matrices, etc..

So it is not at all clear that using either of the above tricks would actually be better than what Flint does already.

However, perhaps there is a case to be made that if Flint is beyond double precision LLL then restarting with lower dimension will be a big saving. There would be quite a lot of code that would have to be written to do this though.

**Comment by wbhart** [comment]
Not reducing dimension might affect S10. Not removing true factors and rebuilding might affect H2.

But doing either might screw up all the other really good timings we have!

**Comment by wbhart** [comment]
Actually, I think Flint uses trace truncation (feeding the top 50 bits or so into LLL, along with all the attendant difficulties of now working with approximations) to keep LLL precision low. So probably, detecting use of mpf_lll would be so rare that it would not help. I guess we could check with the profiler if H2 is somehow triggering multiple expensive full precision LLLs. If so, reducing the dimension with one of the standard tricks, the first time this happens, could be a solution.

This trace truncation thing was the idea of the original ULLL algorithm that Andy came up with, I think. This was hoped to reduce the LLL time to quasilinear in bit size. (From memory some people including Andy, eventually published such an algorithm, but it relied on some additional ideas such as lift-reduction.) But guaranteeing quasilinear behaviour in all cases, and making use of it opportunistically as Flint's polynomial factoring does, are two different things. One is theoretical, the other practical.

**Comment by fredrik-johansson** [comment]
With #1715, #1757, #2381, #2385, #2386, #2393, #2396, #2698, #2702, #2792 and others done in the last couple of years, we should now be well ahead of FLINT 1.6, Pari, Magma and NTL on most of the benchmark problems. Pari is only slightly faster (within 20%) on a couple of examples including M12_6 and H1.

Still keeping this issue open:

* I'm sure we're still missing some tricks in van Hoeij, leading to more LLL work than needed
* Hensel lifting can still be improved a bit
* We're missing some tricks for special-form input (generalize power hack to arbitrary decompositions; checking for cyclotomic factors; special-form irreducibility tests...)
* We might want further fine-tuning for a larger benchmark suite
